From e048f4fbfb52e686c62f464306ca8b5a458a7530 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Wed, 20 Nov 2024 15:58:35 +0800
Subject: [PATCH 50/98] compositor: Early buffer attach for performance

Move renderer attach from repaint to surface attach to reduce latency.

Fix:
c08a6ff8b compositor/renderer: Only attach buffer to renderer in repaint

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 libweston/surface-state.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/libweston/surface-state.c b/libweston/surface-state.c
index 59643dc7..7a5a8f21 100644
--- a/libweston/surface-state.c
+++ b/libweston/surface-state.c
@@ -273,6 +273,14 @@ weston_surface_attach(struct weston_surface *surface,
 	weston_buffer_reference(&surface->buffer_ref, buffer,
 				BUFFER_MAY_BE_ACCESSED);
 
+	/* Early attach to reduce repaint latency */
+	if (!wl_list_empty(&surface->paint_node_list)) {
+		struct weston_paint_node *pnode =
+			wl_container_of(surface->paint_node_list.next,
+					pnode, surface_link);
+		surface->compositor->renderer->attach(pnode);
+	}
+
 	return status;
 }
 
-- 
2.39.5

